### PR TITLE
fix(volta): make sure volta uses the same yarn package overwrite

### PR DESCRIPTION
**Pull request by pataar**
## Changes

This makes sure that the 'volta' behavior for Yarn is the same as the 'engines' behavior.

## Context

When using yarn 2+ in Volta, Renovate will make rollback PR's because it can't find new versions. This is caused because it's looking in the wrong repository.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
please revert the format changes

**Review by viceice — COMMENTED** [review]
i think we already set his anywhere else.

this also doesn't work for simple ranges like `^3.2.0`

**Comment by pataar** [comment]
> i think we already set his anywhere else.
> 
> this also doesn't work for simple ranges like `^3.2.0`

Yes, it's the same piece of code as used in the check for engines:
https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/npm/extract/index.ts#L203

Ranges are not supported in Volta, so that shouldn't be an issue.

**Comment by viceice** [comment]
ok, but it will fill soon again as yarn V4 is around the corner.

so i think you should check if it's not starting with `1.`

**Comment by pataar** [comment]
> ok, but it will fill soon again as yarn V4 is around the corner.
> 
> so i think you should check if it's not starting with `1.`

I've improved the check for both cases

File: lib/modules/manager/npm/extract/index.spec.ts [hunk 1]
```diff
@@ -483,6 +483,48 @@ describe('modules/manager/npm/extract/index', () => {
...
+          yarn: '3.2.4',
+        },
+      };
+      const pJsonStr = JSON.stringify(pJson);
+      const res = await npmExtract.extractPackageFile(
+        pJsonStr,
+        'package.json',
+        defaultConfig
+      );
+      expect(res).toMatchSnapshot({
```
Comment (viceice) [inline lib/modules/manager/npm/extract/index.spec.ts:503]:
```suggestion
      expect(res).toMatchObject({
```

File: lib/modules/manager/npm/extract/index.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import { major } from '@renovatebot/ruby-semver';
```
Comment (viceice) [inline lib/modules/manager/npm/extract/index.ts:1]: ? 😉

Comment (pataar) [reply lib/modules/manager/npm/extract/index.ts]: Sorry, vscode imported this automatically

File: lib/modules/manager/npm/extract/index.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import is from '@sindresorhus/is';
+import semver from 'semver';
```
Comment (viceice) [inline lib/modules/manager/npm/extract/index.ts:2]: use npm versioning

File: lib/modules/manager/npm/extract/index.ts [hunk 4]
```diff
@@ -201,8 +202,8 @@ export async function extractPackageFile(
         dep.commitMessageTopic = 'Yarn';
         constraints.yarn = dep.currentValue;
         if (
-          dep.currentValue.startsWith('2') ||
-          dep.currentValue.startsWith('3')
+          semver.valid(dep.currentValue) &&
+          semver.major(dep.currentValue) > 1
```
Comment (viceice) [inline lib/modules/manager/npm/extract/index.ts:206]: @rarkins we should think about how to support ranges. 🤔

Comment (pataar) [reply lib/modules/manager/npm/extract/index.ts:206]: I'm pretty sure `semver.major("^3.0.0")` would output '3' right? Do you have an example where this wouldn't work? Perhaps it needs an other solution

File: lib/modules/manager/npm/extract/index.spec.ts [hunk 5]
```diff
@@ -483,6 +483,48 @@ describe('modules/manager/npm/extract/index', () => {
...
+            depType: 'volta',
+            packageName: 'nodejs/node',
+            prettyDepType: 'volta',
+            versioning: 'node',
+          },
+          {
+            commitMessageTopic: 'Yarn',
+            currentValue: '3.2.4',
+            datasource: 'npm',
+            depName: 'yarn',
```
Comment (viceice) [inline lib/modules/manager/npm/extract/index.spec.ts:520]: you forgot packageName to validate

Comment (pataar) [reply lib/modules/manager/npm/extract/index.spec.ts:522]: Good one! Added it.

**Review by viceice — APPROVED** [review]

**Pull request merged**